dask/dataframe/core.py

```diff
@@ -4204,17 +4204,37 @@ def _take_last(a, skipna=True):
         Whether to exclude NaN
 
     """
+
+    def _collect_series_last_valid(s):
+        size = len(s)
+        for i in range(size - 1, max(size - 11, 0) - 1, -1):
+            val = s.iloc[i]
+            if not pd.isnull(val):
+                return val
+        else:
+            nonnull = s[s.notna()]
+            if not nonnull.empty:
+                return nonnull.iloc[-1]
+        return None
+
+    def _collect_frame_last_valid(frame):
+        last_valid_rows = {}
+        for col in a.columns:
+            last_val = _collect_series_last_valid(a[col])
+            last_valid_rows[col] = last_val
+        return last_valid_rows
```

Comment:
To reduce indirection, I recommend that we replace this function definition with a dictionary comprehension lower down.